### PR TITLE
Show when an integration is being setup in the UI

### PR DESCRIPTION
**Pull request by bdraco**
## Proposed change

Screen Shot 2022-08-22 at 4 18 04 PM
Screen Shot 2022-08-22 at 4 18 10 PM

It can sometimes be difficult to figure out which integrations are still being setup at startup or why an entity is not available. If the user has 12 config entries for integration X and integration X is blocking startup, which entry is it?

This makes it obvious which integrations are still being setup.

This change also solves a major point of confusion in the UI where it can appear that an integration is in a good state but it's really retrying setup to recover from a failed state and may go back to a failed retry state.

This state was only added recently so it was not possible to know this before, and it wasn't until 2022.8 that we fixed the deadlock that allowed config entries to await their platforms

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/translations/en.json [hunk 1]
```diff
@@ -2837,6 +2837,7 @@
             "area": "In {area}",
             "no_area": "No Area",
             "not_loaded": "Not loaded",
+            "setup_in_progress": "Setup in progress",
```
Comment (balloob) [inline src/translations/en.json:2840]: Can we make this "Initializing". Setup can be confusing.
```suggestion
            "setup_in_progress": "Initializing",
```

File: src/translations/en.json [hunk 2]
```diff
@@ -2857,7 +2858,8 @@
               "migration_error": "Migration error",
               "setup_retry": "Retrying setup",
               "not_loaded": "Not loaded",
-              "failed_unload": "Failed to unload"
+              "failed_unload": "Failed to unload",
+              "setup_in_progress": "Setup in progress"
```
Comment (balloob) [inline src/translations/en.json:2862]:
```suggestion
              "setup_in_progress": "Initializing"
```

**Review by balloob — DISMISSED** [review]
Change looks good.

One of the things I noticed the other day is that we don't actually subscribe to changes to config entry state, causing users to refresh the page to see updates.

**Comment by matthiasdebaat** [comment]
Can we use one of these circular progress indicator?
CleanShot 2022-08-24 at 10 16 49@2x
https://material.io/components/progress-indicators#circular-progress-indicators

**Comment by bdraco** [comment]
> Can we use one of these circular progress indicator? CleanShot 2022-08-24 at 10 16 49@2x [material.io/components/progress-indicators#circular-progress-indicators](https://material.io/components/progress-indicators#circular-progress-indicators)

I tried adding animation at first but its pretty overwhelming at startup when there are lots of integrations loading so I went with the static icon.

**Comment by bdraco** [comment]
> One of the things I noticed the other day is that we don't actually subscribe to changes to config entry state, causing users to refresh the page to see updates.

I don't think there is a core api for this unless I missed it?

**Comment by bdraco** [comment]
I dug through the code and I don't see anything for a stream of config entry state changes.

We would need a subscription that sent, add, delete, changes

**Comment by bramkragten** [comment]
Don't we fire events when a config entry changes?

**Comment by bdraco** [comment]
> Don't we fire events when a config entry changes?

Not that I can see.  There is no machinery to notify when a config entry has been added, removed, or state changes.

We just change the variable
```
config_entries.py:_ConfigEntryStateSelfT = TypeVar("_ConfigEntryStateSelfT", bound="ConfigEntryState")
config_entries.py:class ConfigEntryState(Enum):
config_entries.py:        cls: type[_ConfigEntryStateSelfT], value: str, recoverable: bool
config_entries.py:    ) -> _ConfigEntryStateSelfT:
config_entries.py:        """Create new ConfigEntryState."""
config_entries.py:        state: ConfigEntryState = ConfigEntryState.NOT_LOADED,
config_entries.py:            self.state = ConfigEntryState.SETUP_IN_PROGRESS
config_entries.py:                self.state = ConfigEntryState.SETUP_ERROR
config_entries.py:                self.state = ConfigEntryState.SETUP_ERROR
config_entries.py:                self.state = ConfigEntryState.MIGRATION_ERROR
config_entries.py:            self.state = ConfigEntryState.SETUP_RETRY
config_entries.py:            self.state = ConfigEntryState.LOADED
config_entries.py:            self.state = ConfigEntryState.SETUP_ERROR
config_entries.py:            self.state = ConfigEntryState.NOT_LOADED
config_entries.py:        if self.state == ConfigEntryState.NOT_LOADED:
config_entries.py:                self.state = ConfigEntryState.NOT_LOADED
config_entries.py:            if self.state is not ConfigEntryState.LOADED:
config_entries.py:                self.state = ConfigEntryState.NOT_LOADED
config_entries.py:                self.state = ConfigEntryState.FAILED_UNLOAD
config_entries.py:                self.state = ConfigEntryState.NOT_LOADED
config_entries.py:                self.state = ConfigEntryState.FAILED_UNLOAD
config_entries.py:            unload_success = entry.state is not ConfigEntryState.FAILED_UNLOAD
config_entries.py:        if entry.state is not ConfigEntryState.NOT_LOADED:
config_entries.py:        return entry.state is ConfigEntryState.LOADED  # type: ignore[comparison-overlap] # mypy bug?
config_entries.py:                in (ConfigEntryState.LOADED, ConfigEntryState.SETUP_RETRY)
config_entries.py:                and entry.state is ConfigEntryState.SETUP_RETRY
```

**Comment by balloob** [comment]
Yeah it's something we should fix in another PR.

**Review by balloob — APPROVED** [review]
I am going to merge this but we should re-visit if we can improve the way we show it's in progress.

**Comment by bdraco** [comment]
> I am going to merge this but we should re-visit if we can improve the way we show it's in progress.

https://github.com/home-assistant/frontend/pull/13585

**Pull request merged**